### PR TITLE
Add new Jinja filters

### LINKED ISSUE #61502: [FEATURE REQUEST] Add Jinja filters for itertools functions and common workflows
**Describe the solution you'd like**
I'd love access to some `itertools` functions, a list flattener, and a filter to turn dictionaries into Salt state parameters (a list of single-key dictionaries).

**Describe alternatives you've considered**
Jinja `for` loops mostly.

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
Adds access to some `itertools` functions, a list flattener, and a filter to turn dictionaries into Salt state parameters (a list of single-key dictionaries).

### What issues does this PR fix or reference?
Fixes: #61502

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by nicholasmhughes** [comment]
re-run pr-debian-9-amd64-py3-pytest

**Review by garethgreenaway — DISMISSED** [review]

File: salt/utils/data.py [hunk 1]
```diff
@@ -1541,3 +1541,61 @@ def _append_placeholder(value_dict, key):
         else:
             return [{"value": default if obj is not None else obj}]
     return res
+
+
+@jinja_filter("flatten")
+def flatten(data, levels=None, preserve_nulls=False):
```
Comment (waynew) [inline salt/utils/data.py:1547]: **question** Should this try to detect reference cycles? e.g. with:

```
x = [1,2,3]
x.append(x)
```

Trying to flatten x will currently fail with a RecursionError. Is that correct behavior, or do we want a different behavior?

**Review by waynew — COMMENTED** [review]
Generally I'm +1 here - just one question about recursion...

File: salt/utils/data.py (hunk 1 above)
Comment (nicholasmhughes) [reply salt/utils/data.py:1547]: hmmm... good point. This was lifted pretty faithfully from an existing [Ansible filter](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/filter/core.py#L455), so I didn't put a whole lot of thought into the logic.

@waynew , do you have a suggestion for how to guard against that?

Comment (cmcmarrow) [reply salt/utils/data.py:1547]:
```
def flatten(data, levels=None, preserve_nulls=False, _ids = None):
    if _ids is None:
        _ids = set()
   if id(data) in set():
         "refence circle!" # raise error or return something like [f"Ref Circle"]

   _ids.add(id(data))
   ...
                 ...  = flatten(data, levels, preserve_nulls, _ids):
   ...
```

**Review by garethgreenaway — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Comment by Ch3LL** [comment]
@waynew can you re-review please?

**Pull request merged**